compiler/typecheck/TcErrors.hs

```diff
@@ -2404,6 +2405,28 @@ mkDictErr ctxt cts
     -- but we really only want to report the latter
     elim_superclasses cts = mkMinimalBySCs ctPred cts
 
+customDamlErrors :: Ct -> [ClsInst] -> SDoc -> Maybe SDoc
+customDamlErrors ct candidate_insts binds_msg
+  | TyConApp con [TyConApp target [], viewType] <- ctev_pred (ctEvidence ct)
+  , "HasInterfaceView" <- occNameString $ occName $ tyConName con
+  = Just
+  $ vcat [ text "Tried to get an interface view of type" <+> ppr viewType <+> text "from a non-interface" <+> ppr (tyConName target)
+         , text "If" <+> ppr (tyConName target) <+> text "is a template, try casting it using toInterface or toInterfaceContractId"
+         ]
+  | TyConApp con [TyConApp target [], TyConApp choiceName [], result] <- ctev_pred (ctEvidence ct)
+  , "HasExercise" <- occNameString $ occName $ tyConName con
+  = Just
+  $ vcat [ text "Tried to exercise a choice" <+> ppr choiceName <+> text "which doesn't exist on" <+> ppr (tyConName target)
+         , text "If the choice" <+> ppr choiceName <+> text "belongs to an interface, try casting" <+> ppr (tyConName target) <+> text "using toInterface or toInterfaceContractId"
```

Comment:
Taking this a bit further, if we allow `customDamlErrors` to work in `TcM`, we can use `tcLookupDataCon` on `tyConName target` to check the constructor's "stupid theta" and give an even more precise error message depending on if it's a template, interface or neither. Again, not a blocker, but I really like where this proposal can take us :)